docs/LoongArch-ELF-ABI-EN.adoc

```diff
@@ -628,3 +628,197 @@ with check 32-bit unsigned overflow
+The argument registers are collectively denoted as *AR* (*GAR* or *FAR*), where the general-purpose argument register is denoted as *GAR* and the floating-point argument register is denoted as *FAR*.
+
+=== ABI LP64D
+
+That is, *GRLEN* = 64, *FRLEN* = 64.
+
+=== C Data Types and Alignment
+
+The C data types and alignment in the LP64D ABI are defined in the table 3.
+
+The unsigned integer data types are zero-extended when stored in general-purpose register, and the signed integer data types are sign-extended.
+
+In the *LP64D* ABI, 32-bit types, such as `*int*`, are stored in general-purpose registers as proper sign extensions of their 32-bit values.
+This restriction holds even for unsigned 32-bit types.
+
+=== Argument passing
+
+Generally speaking, FARs are only used to pass floating-point arguments, GARs are used to pass non floating-point arguments and floating-point arguments when no FAR is available(long double type is also passed in a pair of GARs) and the reference.
+
+Arguments passed by reference may be modified by the callee.
+
+==== Scalar
+
+There are three cases:
+
+. 0 < WOA ≤ GRLEN
+
+.. Argument is passed in a single AR, or on the stack by value if none is available.
+
+... If the argument is float-point type, the argument is passed in FAR.
+if no FAR is available, it’s passed in GAR. If no GAR is available, it’s passed on the stack.
+When passed in registers or on the stack, floating-point types narrower than GRLEN bits are widened to GRLEN bits, with the upper bits undefined.
+
+... If the argument is integer or pointer type, the argument is passed in GAR.
+If no GAR is available, it’s passed on the stack.
+When passed in registers or on the stack, integer scalars narrower than GRLEN bits are sign-extended to GRLEN bits.
+
+. GRLEN < WOA ≤ 2 × GRLEN
+
+.. The argument is passed in a pair of GAR, with the low-order GRLEN bits in the lower-numbered register and the high-order GRLEN bits in the higher-numbered register.
+If exactly one register is available, the low-order GRLEN bits are passed in the register and the high-order GRLEN bits are passed on the stack.
+If no GAR is available, it’s passed on the stack.
+
+. WOA > 2 × GRLEN
+
+.. Scalars are passed by reference and are replaced in the argument list with the address.
+
+==== Structure
+
+Empty structures are ignored by C compilers which support them as a non-standard extension(same as union arguments and return values).
+Bits unused due to padding, and bits past the end of a structure whose size in bits is not divisible by GRLEN, are undefined.
+And the layout of the structure on the stack is consistent with that in memory.
+
+. 0 < WOA ≤ GRLEN
+
+.. The structure has only fixed-point members.
+If there is an available GAR, the structure is passed through the GAR by value passing; If no GAR is available, it’s passed on the stack.
```

Comment:
This means that integers and trivial struct wrappers of integers are not passed the same way. Integers are sign-extented; trivial struct wrappers of integers are not sign-extended (upper bits are undefined).

Is this example correct?
```
void A(int x); // x is passed in `a0` register, upper 32-bit of `a0` are sign-extended

struct S
{
    int f;
};

void B(struct S s); // s is passed in `a0` registers, upper 32-bit of `a0` registers are undefined
```

Reply:
> This means that integers and trivial struct wrappers of integers are not passed the same way. Integers are sign-extented; trivial struct wrappers of integers are not sign-extended (upper bits are undefined).
> 
> Is this example correct?
> 
> ```
> void A(int x); // x is passed in `a0` register, upper 32-bit of `a0` are sign-extended
> 
> struct S
> {
>     int f;
> };
> 
> void B(struct S s); // s is passed in `a0` registers, upper 32-bit of `a0` registers are undefined
> ```

There is only one field, the upper 32-bits should be sign extened.

If  `struct S {char  a; cha b;}` , may be just passed by value, the upper 32-bits are undefined.

Reply:
This sounds contrary to https://github.com/loongson/LoongArch-Documentation/pull/32/files#diff-711b3e7b6a005b492898ac6d93f2d8d37c00e0831e210993a6f9dbb26c043717R699 : Bits unused due to padding, and bits past the end of a structure whose size in bits is not divisible by GRLEN, are undefined.